### DISCUSSION TITLE
NLP problem formulation with gradient and Jacobian

### DISCUSSION THREAD

**Discussion by sixtosevenf** [casadi/casadi#3429] [Q&A]
Hi,

I am trying to use the gradient of cost function and the Jacobian of constraints information for an NLP problem.

An optimal control problem (OCP) is transcribed into an NLP problem with the multiple shooting method formulation, i.e., the system state and control input discretized into segments $x \in \mathbb{R}^{n_x \times (N+1)}$, $u \in \mathbb{R}^{n_u \times N}$ where $N$ is the OCP problem horizon. Also, the cost function is $\sum_{N} x^{\top}Qx + u^{\top}Ru$ and the constraint $g(x,u) \leq 0$.

For this problem, where we have N+1 optimization variable for the state and N optimization variable for the control, how can we formulate the gradient and Jacobian for multiple shooting NLP? (I am thinking that the gradient and Jacobian are described with respect to each $x_k$ and $u_k$ but I have not tried this before, so any tip would be helpful.)

Any suggestion would be appreciated.
Thanks

**Comment by sandeep026** [comment]
This explained in the blog post towards the end.  Casadi does the calculation for derivatives behind the scene using AD.  You just have to formulate the NLP

https://web.casadi.org/blog/ocp/

**Reply by sixtosevenf** [reply]
Thanks for your reply. Yes, AD works very well in general cases.

In my problem, the OCP is formulated as bi-level optimization problem and the gradient can be **specially** calculated in an analytical way. I am guessing the analytical gradient would be more computationally efficient than AD.

I would like to compare the computational complexity between the analytical gradient (for "the special case") and AD but I am struggling with formulating the gradient and Jacobian for multiple shooting framework.

**Reply by jaeandersson** [reply]
AD is analytical calculation. It's the chain rule applied to each element of the computational graph. Gradients in particular are calculated using the reverse mode of algorithmic differentiation and the complexity is generally a small factor larger than evaluating the original function, regardless of dimension.

That being said, there are ways to provide specific derivative functions to be used instead of the automatically created ones. It's sometimes used to provide simplified derivatives (e.g. Hessian approximations for least-squares problems) or to reuse derivatives from a previous instance. You can also extract the automatically generated derivative functions from the NLP instance and compare the evaluation time with something you created manually.

**Reply by sixtosevenf** [reply]
Thank Joel for your reply.

I understood the syntax related to Jacobian and gradient now. In your suggestion,  

> You can also extract the automatically generated derivative functions from the NLP instance

Could you give some more details about how to extract the generated derivative functions from NLP result?
